### PR TITLE
Not send wovn-ignore or script tag to Translation API 

### PR DESCRIPTION
**Pull request by mrasu**
### Purpose/goal of PR (Issue #)
Not send wovn-ignore or script tag to Translation API to cache faster and not send secure information.

### Release comments (new feature)
* Add snippet to ensure always snippet is inserted.
* Remove wovn-ignore element and children
* Remove Form to avoid CSRF
* Remove Script to avoid CSRF
* Remove content surrounded by "backend-wovn-ignore" comment

### CODE REVIEW

File: src/wovnio/html/HtmlConverter.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+   *
+   * @return array converted html and HtmlReplaceMarker
+   */
+  public function convertToAppropriateForApiBody() {
+    // encoding detection uses 30% of execution time for this method, should we make encoding as setting?
+    $encoding = mb_detect_encoding($this->html, array('UTF-8', 'EUC-JP', 'SJIS', 'eucJP-win', 'SJIS-win', 'JIS', 'ISO-2022-JP', 'ASCII'));
+    $dom = str_get_html($this->html, false, false, $encoding, false);
+    $marker = new HtmlReplaceMarker();
+    $this->insertSnippet($dom);
+    $this->removeWovnIgnore($dom, $marker);
```
Comment (a-bougouin) [inline src/wovnio/html/HtmlConverter.php]: For Unified Values support, I am not sure we can completely remove `wovn-ignore` content.
@kyuden can you confirm.

File: src/wovnio/html/HtmlConverter.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+  /**
+   * Convert to appropriate HTML to send to Translation API
+   * e.g.) remove wovn-ignore content
+   *
+   * @return array converted html and HtmlReplaceMarker
+   */
+  public function convertToAppropriateForApiBody() {
+    // encoding detection uses 30% of execution time for this method, should we make encoding as setting?
+    $encoding = mb_detect_encoding($this->html, array('UTF-8', 'EUC-JP', 'SJIS', 'eucJP-win', 'SJIS-win', 'JIS', 'ISO-2022-JP', 'ASCII'));
```
Comment (a-bougouin) [inline src/wovnio/html/HtmlConverter.php]: I agree. If user always use same encoding, they should be able to specify settings for faster execution. I user does not specify encoding, we should keep detection.

If we do that, we should also document special setup for optimization purpose (not related bit need to do same with curl library).

File: test/html/HtmlReplaceMakerTest.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?php
+require_once 'src/wovnio/html/HtmlReplaceMarker.php';
+
+use Wovnio\Html\HtmlReplaceMarker;
+
+class HtmlReplaceMarkerTest extends PHPUnit_Framework_TestCase {
```
Comment (a-bougouin) [inline test/html/HtmlReplaceMakerTest.php]: Please make tests with at least 11 keys.
I don't see the problem here, but first wovnphp version had revert issues with keys ending by `1` and `11` (or `2` and `20`, etc.). We need to secure that case for future changes.

File: src/wovnio/wovnphp/API.php [hunk 4]
```diff
@@ -19,18 +20,22 @@ public static function url($store, $headers, $original_content) {
     public static function translate($store, $headers, $original_content) {
       $translated_content = NULL;
       $api_url = self::url($store, $headers, $original_content);
+
+      $token = $store->settings['project_token'];
+      $converter = new HtmlConverter($original_content, $token);
+      list($converted_html, $maker) = $converter->convertToAppropriateForApiBody();
```
Comment (a-bougouin) [inline src/wovnio/wovnphp/API.php]: Typo: `$maker` -> `$marker`

**Comment by a-bougouin** [comment]
> Add snippet to ensure always snippet is inserted

Was it requested? When to expect API to not return HTML with snippet?

If there is a bug with API not returning HTML with snippet, then we need to fix the API.
I am concerned that adding the snippet will make it more difficult to catch error and user may lose backend benefit (SEO) without noticing in a short window of time.

File: src/wovnio/html/HtmlConverter.php (hunk 1 above)
Comment (mrasu) [reply src/wovnio/html/HtmlConverter.php]: @kyuden please check

File: src/wovnio/html/HtmlConverter.php (hunk 2 above)
Comment (mrasu) [reply src/wovnio/html/HtmlConverter.php]: done. `encoding` option is created

File: test/html/HtmlReplaceMakerTest.php (hunk 3 above)
Comment (mrasu) [reply test/html/HtmlReplaceMakerTest.php]: done!

File: src/wovnio/wovnphp/API.php (hunk 4 above)
Comment (mrasu) [reply src/wovnio/wovnphp/API.php]: done!

**Comment by mrasu** [comment]
@a-bougouin 
For when WOVN.php cannot reach translation API.

In that case, "j.wovn.io" is still working because "j.wovn.io" doesn't depend on API.

File: src/wovnio/html/HtmlConverter.php (hunk 2 above)
Comment (a-bougouin) [reply src/wovnio/html/HtmlConverter.php]: thanks

File: test/html/HtmlReplaceMakerTest.php (hunk 3 above)
Comment (a-bougouin) [reply test/html/HtmlReplaceMakerTest.php]: thanks

**Comment by a-bougouin** [comment]
@mrasu, then could we insert the snippet after API request?
If snippet exists, do nothing. If snippet does not exists, then add snippet and trigger a Rollbar error. I want us to be notified somehow that snippet was not inserted, so that we can investigate and fix the problem.

What do you think?

**Comment by mrasu** [comment]
@a-bougouin seems nice. I added "data-wovnio-type" to snippet

**Comment by mrasu** [comment]
@a-bougouin  :+1:

File: src/wovnio/html/HtmlConverter.php (hunk 1 above)
Comment (kyuden) [reply src/wovnio/html/HtmlConverter.php]: In current implementation of swapper, wovn-ignore is needed. original content is key to find the destination.

Comment (kyuden) [reply src/wovnio/html/HtmlConverter.php]: I talk with @mrasu in person. we don't need wovn-ignore content. but wovn-ignore tag is needed.

Comment (mrasu) [reply src/wovnio/html/HtmlConverter.php]: ok done.

**Comment by a-bougouin** [comment]
Tests need to be fixed.

@WOVNio/developers can someone review please.

File: test/APITest.php [hunk 5]
```diff
@@ -129,19 +140,54 @@ public function testTranslateWithCustomLangAliases() {
       $result = API::translate($store, $headers, $html);
       $this->assertEquals($expected_result, $result);
     }
+/*
```
Comment (a-bougouin) [inline test/APITest.php]: Why must comment test?

Comment (mrasu) [reply test/APITest.php]: thanks, removed

**Pull request merged**